File: app/src/main/java/com/digicoffer/lauditor/Meetings/MonthCalander/SingleMonthFragment.java
package com.digicoffer.lauditor.Meetings.MonthCalander;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.digicoffer.lauditor.Appointments.Models.AppointmentModel;
import com.digicoffer.lauditor.Meetings.ViewModels.Meetings;
import com.digicoffer.lauditor.Meetings.Models.Events_Do;
import com.digicoffer.lauditor.R;

import java.time.LocalDate;
import java.time.YearMonth;
import java.util.ArrayList;
import java.util.List;

public class SingleMonthFragment extends Fragment {
    private static final String ARG_DATE = "month_date";
    private LocalDate month;
    private CalendarDayAdapter adapter;
    private CalendarDayAdapter.OnDateSelectedListener dateSelectedListener;
    private ArrayList<Events_Do> events_list = new ArrayList<>();
    private ArrayList<AppointmentModel> appointments_list = new ArrayList<>();
    private static Meetings.FilterType currentFilter = Meetings.FilterType.ALL;

    public static SingleMonthFragment newInstance(LocalDate date,
                                                  ArrayList<Events_Do> events_list,
                                                  ArrayList<AppointmentModel> appointments_list, Meetings.FilterType filterType,
                                                  CalendarDayAdapter.OnDateSelectedListener dateSelectedListener) {
        SingleMonthFragment fragment = new SingleMonthFragment();
        Bundle args = new Bundle();
        args.putString(ARG_DATE, date.toString());
        args.putSerializable("currentFilter", filterType);
        args.putSerializable("events_list", events_list);
        args.putSerializable("appointments_list", appointments_list);
        fragment.setArguments(args);
        fragment.setDateSelectedListener(dateSelectedListener);
        return fragment;
    }

    public void setDateSelectedListener(CalendarDayAdapter.OnDateSelectedListener listener) {
        this.dateSelectedListener = listener;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            month = LocalDate.parse(getArguments().getString(ARG_DATE));
            currentFilter = (Meetings.FilterType)
                    getArguments().getSerializable("currentFilter");
            events_list = (ArrayList<Events_Do>) getArguments().getSerializable("events_list");
            appointments_list = (ArrayList<AppointmentModel>) getArguments().getSerializable("appointments_list");

            // Handle null cases
            if (events_list == null) {
                events_list = new ArrayList<>();
            }
            if (appointments_list == null) {
                appointments_list = new ArrayList<>();
            }
        }
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_single_month, container, false);
        RecyclerView recyclerView = view.findViewById(R.id.recyclerView);

        recyclerView.setLayoutManager(new GridLayoutManager(getContext(), 7));
        List<CalendarItem> days = generateCalendarItems(month);

        // Pass both events and appointments to the adapter
        CalendarDayAdapter adapter = new CalendarDayAdapter(days, events_list, appointments_list, currentFilter);
        adapter.setOnDateSelectedListener(dateSelectedListener);
        recyclerView.setAdapter(adapter);

        return view;
    }
    public void updateFilter(Meetings.FilterType newFilter) {
        currentFilter = newFilter;
        if (adapter != null) {
            adapter.updateFilter(newFilter);
        }
    }
    private List<CalendarItem> generateCalendarItems(LocalDate date) {
        List<CalendarItem> list = new ArrayList<>();

        // Add weekday headers
        String[] dayNames = {"Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"};
        for (String day : dayNames) {
            list.add(new CalendarItem(day));
        }

        // Add date placeholders
        YearMonth yearMonth = YearMonth.from(date);
        LocalDate firstOfMonth = yearMonth.atDay(1);
        int startDay = firstOfMonth.getDayOfWeek().getValue() % 7; // Sunday = 0

        // Add empty cells before first day of month
        for (int i = 0; i < startDay; i++) {
            list.add(new CalendarItem((LocalDate) null));
        }

        // Add all days of the month
        for (int i = 1; i <= yearMonth.lengthOfMonth(); i++) {
            list.add(new CalendarItem(yearMonth.atDay(i)));
        }

        // Pad remaining cells to complete the grid (7 columns × 7 rows = 49 cells including headers)
        while (list.size() < 49) {
            list.add(new CalendarItem((LocalDate) null));
        }

        return list;
    }
}

File: app/src/main/java/com/digicoffer/lauditor/Meetings/MonthCalander/SingleMonthFragment.kt
package com.digicoffer.lauditor.Meetings.MonthCalander

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.digicoffer.lauditor.Appointments.Models.AppointmentModel
import com.digicoffer.lauditor.Meetings.ViewModels.Meetings
import com.digicoffer.lauditor.Meetings.Models.Events_Do
import com.digicoffer.lauditor.R
import java.time.LocalDate
import java.time.YearMonth
import java.util.ArrayList

class SingleMonthFragment : Fragment() {
    private var month: LocalDate? = null
    private var adapter: CalendarDayAdapter? = null
    private var dateSelectedListener: CalendarDayAdapter.OnDateSelectedListener? = null
    private var events_list = ArrayList<Events_Do>()
    private var appointments_list = ArrayList<AppointmentModel>()

    fun setDateSelectedListener(listener: CalendarDayAdapter.OnDateSelectedListener?) {
        this.dateSelectedListener = listener
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        arguments?.let { args ->
            val dateStr = args.getString(ARG_DATE)
            if (dateStr != null) {
                month = LocalDate.parse(dateStr)
            }
            val filter = args.getSerializable("currentFilter")
            if (filter is Meetings.FilterType) {
                currentFilter = filter
            }
            val evList = args.getSerializable("events_list")
            if (evList is ArrayList<*>) {
                events_list = evList as ArrayList<Events_Do>
            }
            val appList = args.getSerializable("appointments_list")
            if (appList is ArrayList<*>) {
                appointments_list = appList as ArrayList<AppointmentModel>
            }
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_single_month, container, false)
        val recyclerView = view.findViewById<RecyclerView>(R.id.recyclerView)

        recyclerView.layoutManager = GridLayoutManager(context, 7)
        val days = generateCalendarItems(month ?: LocalDate.now())

        // Pass both events and appointments to the adapter
        adapter = CalendarDayAdapter(days, events_list, appointments_list, currentFilter)
        adapter?.setOnDateSelectedListener(dateSelectedListener)
        recyclerView.adapter = adapter

        return view
    }

    fun updateFilter(newFilter: Meetings.FilterType) {
        currentFilter = newFilter
        adapter?.updateFilter(newFilter)
    }

    private fun generateCalendarItems(date: LocalDate): List<CalendarItem> {
        val list = ArrayList<CalendarItem>()

        // Add weekday headers
        val dayNames = arrayOf("Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat")
        for (day in dayNames) {
            list.add(CalendarItem(day))
        }

        // Add date placeholders
        val yearMonth = YearMonth.from(date)
        val firstOfMonth = yearMonth.atDay(1)
        val startDay = firstOfMonth.dayOfWeek.value % 7 // Sunday = 0

        // Add empty cells before first day of month
        for (i in 0 until startDay) {
            list.add(CalendarItem(null as LocalDate?))
        }

        // Add all days of the month
        for (i in 1..yearMonth.lengthOfMonth()) {
            list.add(CalendarItem(yearMonth.atDay(i)))
        }

        // Pad remaining cells to complete the grid (7 columns x 7 rows = 49 cells including headers)
        while (list.size < 49) {
            list.add(CalendarItem(null as LocalDate?))
        }

        return list
    }

    companion object {
        private const val ARG_DATE = "month_date"
        private var currentFilter = Meetings.FilterType.ALL

        @JvmStatic
        fun newInstance(
            date: LocalDate,
            events_list: ArrayList<Events_Do>,
            appointments_list: ArrayList<AppointmentModel>,
            filterType: Meetings.FilterType,
            dateSelectedListener: CalendarDayAdapter.OnDateSelectedListener
        ): SingleMonthFragment {
            val fragment = SingleMonthFragment()
            val args = Bundle().apply {
                putString(ARG_DATE, date.toString())
                putSerializable("currentFilter", filterType)
                putSerializable("events_list", events_list)
                putSerializable("appointments_list", appointments_list)
            }
            fragment.arguments = args
            fragment.setDateSelectedListener(dateSelectedListener)
            return fragment
        }
    }
}
